internal/report/write.go

```diff
@@ -0,0 +1,81 @@
+package report
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	polReport "sigs.k8s.io/wg-policy-prototypes/policy-report/pkg/api/wgpolicyk8s.io/v1beta1"
+)
+
+// Write instantiates the passed namespaced PolicyReport if it doesn't exist, or
+// updated a new one if one is found
+func Write(report *polReport.PolicyReport) error {
+	config := ctrl.GetConfigOrDie()
+	customScheme := scheme.Scheme
+	customScheme.AddKnownTypes(
+		polReport.SchemeGroupVersion,
+		&polReport.PolicyReport{},
+		&polReport.PolicyReportList{},
+		&polReport.ClusterPolicyReportList{},
+	)
+	metav1.AddToGroupVersion(customScheme, polReport.SchemeGroupVersion)
+	client, err := client.New(config, client.Options{Scheme: customScheme})
+	if err != nil {
+		return fmt.Errorf("failed when creating new client: %w", err)
+	}
+
+	// Check for existing Policy Reports
+	result := &polReport.PolicyReport{}
+	getErr := client.Get(context.TODO(), types.NamespacedName{
+		Namespace: report.Namespace,
+		Name:      report.Name,
+	}, result)
+
+	// Create new Policy Report if not found
+	if errors.IsNotFound(getErr) {
+		log.Info().Msg("creating policy report...")
+
+		err = client.Create(context.TODO(), report)
+		if err != nil {
+			return fmt.Errorf("failed when creating PolicyReport: %w", err)
+		}
+	} else {
+		// Update existing Policy Report
```

Comment:
I wonder if we shouldn't just delete the old report and always create a new one :thinking:

Reply:
I thought that naively deleting the old report without RetryOnConflict could be somewhat dangerous. And once I had the retry, why not just do an update for the full Report, it is going to overwrite it anyways. Plus this way, we bump the version for possible listeners.